### ISSUE TITLE
Logging system

### ISSUE THREAD

**Issue by Sogl** [koala-framework/koala-framework#592]
Hello!

I created log table for my model that works the same as `kwf_user_messages` log system.
It works but now I need to create logging system for all specified models (because I want to exclude all system tables `kwf_*` etc).

Finally in frontend ExtJS part I can see what object\model, what user, IP and what data was changed.

I mean something like that:
![1-eb8c81f-f5q_a_cab01rzw](https://cloud.githubusercontent.com/assets/7556290/22715129/afd98bfe-ede3-11e6-8303-4d0a1b1f4602.png)

I need something like middleware or global `_afterInsert`/`_afterSave`/`_afterDelete` event.

Where I can start?

**Comment by nsams** [comment]
We have an event system that you can use for that.

First, you need to define a event subscriber class, you can do that in a model:
    
    class Logs extends Kwf_Model_Db
    {
        protected $_table = 'logs';
        public function getEventSubscribers()
        {
            $ret = parent::getEventSubscribers();
            $ret[] = Kwf_Model_EventSubscriber::getInstance('Logs_Events', array(
                'modelFactoryConfig' => $this->getFactoryConfig()
            ));
            return $ret;
        }
    }

and then you implement this event subscriber class by defining the events you want to listen to (in getListeners) and implementing the callback methods that will be called on that event:

    class Logs_Events extends Kwf_Events_Subscriber
    {
        public function getListeners()
        {
            return array(
                array(
                    'event' => 'Kwf_Events_Event_Row_Updated',
                    'callback' => 'onUpdated'
                ),
                array(
                    'event' => 'Kwf_Events_Event_Row_Deleted',
                    'callback' => 'onDeleted'
                ),
                array(
                    'event' => 'Kwf_Events_Event_Row_Inserted',
                    'callback' => 'onInserted'
                ),
            );
        }
    
        public function onUpdated(Kwf_Events_Event_Row_Updated $ev)
        {
            $ev->row;
            $ev->row->isDirty('foo');
        }
        
        public function onDeleted(Kwf_Events_Event_Row_Updated $ev)
        {
        }
        
        public function onInserted(Kwf_Events_Event_Row_Updated $ev)
        {
        }
    }

**Comment by Sogl** [comment]
>First, you need to define a event subscriber class, you can do that in a model:

Ok. I put it into `models` folder.
But what about DB schema for this table? Or it uses `kwf_logs` table?

>and then you implement this event subscriber class

Where I need to put this file?

**Comment by nsams** [comment]
>> First, you need to define a event subscriber class, you can do that in a model:

> Ok. I put it into models folder.
> But what about DB schema for this table? Or it uses kwf_logs table?

yes, you add the event subscribers to the model those events belong to. in your case the logs table.

>> and then you implement this event subscriber class

> Where I need to put this file?

doesn't matter, eg in moels/Logs/Events.php (class name Logs_Events)

**Comment by Sogl** [comment]
@nsams Ok. I did something near that I wanted. Please have a look:
https://gist.github.com/Sogl/2a4de5289a0b459e0f110ba1835d24fa

I found some problems that I don't know how to resolve:
1.. There are some of system generated events that I want to exclude. For example:

![image](https://cloud.githubusercontent.com/assets/7556290/23111608/bcf95eec-f77d-11e6-885c-b6d9f5fc0979.png)

I don't need `Kwf_User_EditModel.proxied` and others. Where I can exclude them?

2.. When I use my function to collect object data, some objects becomes very large:

![image](https://cloud.githubusercontent.com/assets/7556290/23111893/d22f9e82-f77f-11e6-84b5-37fea42ed91b.png)

Maybe I need to put something like this into every model:
```php
protected $log = [
    'id',
    'firstname',
    'middlename',
    //etc
];
```
?

3.. Can I exclude writing `afterSave`/`afterInsert` events like this?
```php
protected function _afterSave(Kwf_Model_Row_Interface $row)
    {
        if ($row->id != NULL) {

            $linkModel = Kwf_Model_Abstract::getInstance('Linkdata');
            $linkSelect = $linkModel->select()->whereEquals('link_id', $row->id);
            $links = $linkModel->getRows($linkSelect);

            if (count($links) > 0) {

                foreach ($links as $link) {
                    $link->name = $row->name;
                    $link->save();
                }

            }
        }
    }
```
It's more automatic creation/updating and I want exclude or flag all of them.

3.. How to get model names for frontend?
4.. And also can I get field names?

**Comment by nsams** [comment]
We don't have a ready solution for your requirements.

3: you have to do that manually

3(2nd :D)+4: You have to add somewhere arrays for pretty model and field names.

**Comment by Sogl** [comment]
>We don't have a ready solution for your requirements.

I understand this :D 
That's because I'm writing something that can handle some needed things.

>3(2nd :D)+4: You have to add somewhere arrays for pretty model and field names.

Are there any built in functions like `getModelTranslation` (grid title) or `getFieldTranslation` (that put in `initColumns()`) based on current user language?

**Comment by nsams** [comment]
> Any thoughts how to stop firing events in this case?

That is not possible, you can however ignore events from specific models in your code.

> Are there any built in functions like getModelTranslation (grid title) or getFieldTranslation (that put in initColumns()) based on current user language?

We don't have those strings anywhere...
